File: examples/secondary_development/perl/signal_health.pl
#!/usr/bin/env perl
use strict;
use warnings;
use Getopt::Long qw(GetOptions);
use File::Path qw(make_path);
use File::Spec;
use FindBin qw($RealBin);

my ($fsdb, $signal, $begin, $end, $out_dir);
my $max_rows = 200;
my $min_changes = 1;
my $max_unknown = 0;
my $require_complete = 0;
my $help = 0;
my ($kdebug_override, $json_python_override, $json_helper_override);

GetOptions(
    'fsdb=s'            => \$fsdb,
    'signal=s'          => \$signal,
    'begin|start=s'     => \$begin,
    'end|stop=s'        => \$end,
    'max-rows=i'        => \$max_rows,
    'min-changes=i'     => \$min_changes,
    'max-unknown=i'     => \$max_unknown,
    'require-complete!' => \$require_complete,
    'kdebug-bin=s'      => \$kdebug_override,
    'json-python=s'     => \$json_python_override,
    'json-helper=s'     => \$json_helper_override,
    'out=s'             => \$out_dir,
    'help|h'            => \$help,
) or usage(2);
usage(0) if $help;
usage(2) unless defined $fsdb && defined $signal && defined $begin
    && defined $end && defined $out_dir;
usage(2) if $max_rows < 0 || $min_changes < 0 || $max_unknown < 0;

sub usage {
    my ($rc) = @_;
    print STDERR <<'USAGE';
Usage: signal_health.pl --fsdb FILE --signal NAME --begin TIME --end TIME
                        [--max-rows N] [--min-changes N]
                        [--max-unknown N] [--require-complete]
                        [--kdebug-bin CMD] [--json-python CMD]
                        [--json-helper FILE] --out DIR

The script invokes kdebug, parses signal.scan JSON, and emits conclusion.json.
Resolution: explicit option, environment, KVERIF_HOME/tools, repository, PATH.
USAGE
    exit $rc;
}

sub find_executable {
    my ($command) = @_;
    return unless defined $command && length $command;
    return File::Spec->rel2abs($command) if -f $command && -x $command;
    for my $directory (File::Spec->path()) {
        my $candidate = File::Spec->catfile($directory, $command);
        return File::Spec->rel2abs($candidate) if -f $candidate && -x $candidate;
    }
    return;
}

sub configured_command {
    my ($value, $label) = @_;
    my $resolved = find_executable($value);
    defined $resolved or die "$label is not executable: $value\n";
    return $resolved;
}

my $repo_tool = File::Spec->rel2abs(
    File::Spec->catfile($RealBin, '..', '..', '..', 'tools', 'kdebug')
);
my $kdebug;
if (defined $kdebug_override && length $kdebug_override) {
    $kdebug = configured_command($kdebug_override, 'kdebug');
} elsif (defined $ENV{KDEBUG_BIN} && length $ENV{KDEBUG_BIN}) {
    $kdebug = configured_command($ENV{KDEBUG_BIN}, 'kdebug');
} elsif (defined $ENV{KVERIF_HOME}
         && -x File::Spec->catfile($ENV{KVERIF_HOME}, 'tools', 'kdebug')) {
    $kdebug = File::Spec->catfile($ENV{KVERIF_HOME}, 'tools', 'kdebug');
} elsif (-x $repo_tool) {
    $kdebug = $repo_tool;
} else {
    $kdebug = find_executable('kdebug');
}
defined $kdebug or die "cannot find kdebug; use --kdebug-bin, KDEBUG_BIN, KVERIF_HOME, or PATH\n";

my $python_setting = defined $json_python_override && length $json_python_override
    ? $json_python_override
    : defined $ENV{KVERIF_JSON_PYTHON} && length $ENV{KVERIF_JSON_PYTHON}
        ? $ENV{KVERIF_JSON_PYTHON}
        : defined $ENV{PYTHON} && length $ENV{PYTHON}
            ? $ENV{PYTHON}
            : undef;
my $json_python;
if (defined $python_setting) {
    $json_python = configured_command($python_setting, 'JSON Python');
} else {
    $json_python = find_executable('python3') || find_executable('python');
}
defined $json_python or die "cannot find Python 3; use --json-python, KVERIF_JSON_PYTHON, or PYTHON\n";
system($json_python, '-c', 'import json, sys; raise SystemExit(0 if sys.version_info[0] >= 3 else 1)') == 0
    or die "JSON helper requires Python 3: $json_python\n";

my $local_helper = File::Spec->rel2abs(File::Spec->catfile($RealBin, '..', 'json_response.py'));
my $home_helper = defined $ENV{KVERIF_HOME}
    ? File::Spec->catfile($ENV{KVERIF_HOME}, 'examples', 'secondary_development', 'json_response.py')
    : '';
my $json_helper = defined $json_helper_override && length $json_helper_override
    ? $json_helper_override
    : defined $ENV{KVERIF_JSON_HELPER} && length $ENV{KVERIF_JSON_HELPER}
        ? $ENV{KVERIF_JSON_HELPER}
        : -f $local_helper
            ? $local_helper
            : length $home_helper && -f $home_helper
                ? $home_helper
                : $local_helper;
-f $json_helper or die "JSON helper is missing: $json_helper\n";
make_path($out_dir);

my $response = File::Spec->catfile($out_dir, 'tool-response.json');
my $stderr_file = File::Spec->catfile($out_dir, 'tool-response.stderr');
my $report = File::Spec->catfile($out_dir, 'conclusion.json');

sub run_redirected {
    my ($stdout_path, $stderr_path, @command) = @_;
    my $pid = fork();
    defined $pid or die "fork failed: $!\n";
    if ($pid == 0) {
        open STDOUT, '>:encoding(UTF-8)', $stdout_path
            or die "cannot write $stdout_path: $!\n";
        open STDERR, '>:encoding(UTF-8)', $stderr_path
            or die "cannot write $stderr_path: $!\n";
        exec {$command[0]} @command;
        die "cannot exec $command[0]: $!\n";
    }
    waitpid($pid, 0);
    return $? >> 8;
}

sub capture_line {
    my (@command) = @_;
    open my $pipe, '-|', @command or die "cannot start @command: $!\n";
    my $line = <$pipe>;
    close $pipe;
    my $rc = $? >> 8;
    die "command failed rc=$rc: @command\n" if $rc != 0;
    defined $line or die "command produced no output: @command\n";
    chomp $line;
    $line =~ s/\r$//;
    return $line;
}

my @tool_command = (
    $kdebug, '--json', 'action', 'signal.scan',
    '--fsdb', $fsdb,
    '--arg', "signal=$signal",
    '--arg', "begin=$begin",
    '--arg', "end=$end",
    '--arg', 'format=hex',
    '--max-rows', $max_rows,
);
my $tool_rc = run_redirected($response, $stderr_file, @tool_command);
die "kdebug failed rc=$tool_rc; see $stderr_file\n" if $tool_rc != 0;

system($json_python, $json_helper, 'check-ok', $response) == 0
    or die "tool returned ok=false; see $response\n";
my $stats = capture_line($json_python, $json_helper, 'wave-stats', $response);
my ($change_count, $unknown_count, $truncated) = split /\t/, $stats;
defined $truncated or die "invalid wave-stats output: $stats\n";

my $conclusion = 'HEALTHY';
my $reason = 'signal activity satisfies all configured gates';
if ($require_complete && $truncated eq 'true') {
    $conclusion = 'INCOMPLETE';
    $reason = 'signal.scan was truncated, so the complete window was not evaluated';
} elsif ($unknown_count > $max_unknown) {
    $conclusion = 'UNKNOWN_VALUES';
    $reason = 'unknown_count exceeds the configured maximum';
} elsif ($change_count < $min_changes) {
    $conclusion = 'INACTIVE';
    $reason = 'change_count is below the configured minimum';
}

my @report_command = (
    $json_python, $json_helper, 'signal-health-report',
    '--output', $report, '--language', 'perl', '--signal', $signal,
    '--response', $response, '--change-count', $change_count,
    '--unknown-count', $unknown_count, '--truncated', $truncated,
    '--min-changes', $min_changes, '--max-unknown', $max_unknown,
    '--require-complete', $require_complete ? 'true' : 'false',
    '--conclusion', $conclusion, '--reason', $reason,
);
system(@report_command) == 0 or die "failed to write conclusion report\n";

print "signal health: $conclusion ($reason)\n";
print "report: $report\n";
exit($conclusion eq 'HEALTHY' ? 0 : 3);

File: examples/secondary_development/py/signal_health.py
#!/usr/bin/env python3
"""Invoke kdebug as a subprocess and derive a signal-health conclusion."""

from __future__ import print_function

import argparse
import json
import os
import shutil
import subprocess
import sys


def parser():
    result = argparse.ArgumentParser(
        description="Run kdebug signal.scan and derive a project-level conclusion"
    )
    result.add_argument("--fsdb", required=True)
    result.add_argument("--signal", required=True)
    result.add_argument("--begin", "--start", dest="begin", required=True)
    result.add_argument("--end", "--stop", dest="end", required=True)
    result.add_argument("--max-rows", type=int, default=200)
    result.add_argument("--min-changes", type=int, default=1)
    result.add_argument("--max-unknown", type=int, default=0)
    result.add_argument("--require-complete", action="store_true")
    result.add_argument("--kdebug-bin")
    result.add_argument("--out", required=True)
    return result


def integer_summary(summary, name):
    value = summary.get(name, 0)
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        raise ValueError("summary.{0} is not numeric".format(name))
    return int(value)


def classify(change_count, unknown_count, truncated, args):
    if args.require_complete and truncated:
        return (
            "INCOMPLETE",
            "signal.scan was truncated, so the complete window was not evaluated",
        )
    if unknown_count > args.max_unknown:
        return "UNKNOWN_VALUES", "unknown_count exceeds the configured maximum"
    if change_count < args.min_changes:
        return "INACTIVE", "change_count is below the configured minimum"
    return "HEALTHY", "signal activity satisfies all configured gates"


def tool_prefix(path):
    if os.name != "nt":
        return [path]
    try:
        with open(path, "rb") as handle:
            is_script = handle.read(2) == b"#!"
    except OSError:
        is_script = False
    if not is_script:
        return [path]
    bash = os.environ.get("BASH") or shutil.which("bash")
    if not bash:
        raise OSError("bash is required to launch the kdebug wrapper on Windows")
    return [bash, path]


def executable_path(command):
    if not command:
        return None
    expanded = os.path.expanduser(command)
    if os.path.isfile(expanded) and (os.name == "nt" or os.access(expanded, os.X_OK)):
        return os.path.abspath(expanded)
    resolved = shutil.which(command)
    if resolved:
        return resolved
    if os.path.basename(command) == command:
        for directory in os.environ.get("PATH", "").split(os.pathsep):
            candidate = os.path.join(directory, command)
            if os.path.isfile(candidate) and (os.name == "nt" or os.access(candidate, os.X_OK)):
                return os.path.abspath(candidate)
    return None


def resolve_kdebug(explicit):
    configured = explicit or os.environ.get("KDEBUG_BIN")
    if configured:
        resolved = executable_path(configured)
        if not resolved:
            raise OSError("configured kdebug is not executable: {0}".format(configured))
        return resolved

    home = os.environ.get("KVERIF_HOME")
    candidates = []
    if home:
        candidates.append(os.path.join(home, "tools", "kdebug"))
    candidates.append(
        os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "..", "..", "tools", "kdebug"))
    )
    for candidate in candidates:
        resolved = executable_path(candidate)
        if resolved:
            return resolved
    resolved = executable_path("kdebug")
    if resolved:
        return resolved
    raise OSError("cannot find kdebug; use --kdebug-bin, KDEBUG_BIN, KVERIF_HOME, or PATH")


def main(argv=None):
    args = parser().parse_args(argv)
    for name in ("max_rows", "min_changes", "max_unknown"):
        if getattr(args, name) < 0:
            parser().error("--{0} must be non-negative".format(name.replace("_", "-")))

    try:
        kdebug = resolve_kdebug(args.kdebug_bin)
    except OSError as exc:
        print("ERROR: {0}".format(exc), file=sys.stderr)
        return 2

    os.makedirs(args.out, exist_ok=True)
    response_path = os.path.join(args.out, "tool-response.json")
    stderr_path = os.path.join(args.out, "tool-response.stderr")
    report_path = os.path.join(args.out, "conclusion.json")
    command = tool_prefix(kdebug) + [
        "--json",
        "action",
        "signal.scan",
        "--fsdb",
        args.fsdb,
        "--arg",
        "signal={0}".format(args.signal),
        "--arg",
        "begin={0}".format(args.begin),
        "--arg",
        "end={0}".format(args.end),
        "--arg",
        "format=hex",
        "--max-rows",
        str(args.max_rows),
    ]
    try:
        with open(response_path, "w", encoding="utf-8") as stdout_handle, open(
            stderr_path, "w", encoding="utf-8"
        ) as stderr_handle:
            completed = subprocess.run(
                command,
                stdout=stdout_handle,
                stderr=stderr_handle,
                universal_newlines=True,
            )
    except OSError as exc:
        print("ERROR: cannot start kdebug: {0}".format(exc), file=sys.stderr)
        return 1
    if completed.returncode != 0:
        print("ERROR: kdebug failed; see {0}".format(stderr_path), file=sys.stderr)
        return 1

    try:
        with open(response_path, "r", encoding="utf-8") as handle:
            response = json.load(handle)
        if not isinstance(response, dict) or response.get("ok") is not True:
            raise ValueError("tool response is not an ok=true object")
        summary = response.get("summary", {})
        if not isinstance(summary, dict):
            raise ValueError("summary is not an object")
        change_count = integer_summary(summary, "change_count")
        unknown_count = integer_summary(summary, "unknown_count")
        truncated = bool(summary.get("truncated", False))
    except (OSError, ValueError, TypeError) as exc:
        print("ERROR: invalid kdebug JSON: {0}".format(exc), file=sys.stderr)
        return 1

    conclusion, reason = classify(change_count, unknown_count, truncated, args)
    report = {
        "schema": "kverif.example.signal-health.v1",
        "language": "python",
        "gate_pass": conclusion == "HEALTHY",
        "conclusion": {"status": conclusion, "reason": reason},
        "evidence": {
            "signal": args.signal,
            "change_count": change_count,
            "unknown_count": unknown_count,
            "truncated": truncated,
        },
        "thresholds": {
            "min_changes": args.min_changes,
            "max_unknown": args.max_unknown,
            "require_complete": args.require_complete,
        },
        "artifacts": {"tool_response": response_path, "tool_stderr": stderr_path},
    }
    with open(report_path, "w", encoding="utf-8") as handle:
        json.dump(report, handle, ensure_ascii=False, separators=(",", ":"), sort_keys=True)
        handle.write("\n")

    print("signal health: {0} ({1})".format(conclusion, reason))
    print("report: {0}".format(report_path))
    return 0 if conclusion == "HEALTHY" else 3


if __name__ == "__main__":
    sys.exit(main())
